File: cpp_version/exhaustive_recursion/permutations/permutations.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// permutations
// Write a function, permutations, that takes in a list an argument. The function should
// return a 2D list where each sublist represents one of the possible permutations of the list.

// The sublists may be returned in any order.

// You may assume that the input list contains unique items.
void printVector(std::vector<std::vector<std::string>>);
std::vector<std::vector<std::string>> permutations(std::vector<std::string>);
std::vector<std::vector<std::string>> permutations(std::vector<std::string> elements) {
    // todo

    std::vector<std::vector<std::string>> result;
    return result;
}

void printVector(std::vector<std::vector<std::string>> vec) {
    for (auto x : vec) {

        for (auto y : x) {
            std::cout << y << " , ";
        }
    }
    std::cout << std::endl;
}

int main() {

    printVector(permutations(std::vector<std::string>{"a", "b", "c"})); // ->
    // [
    //   [ "a", "b", "c" ],
    //   [ "b", "a", "c" ],
    //   [ "b", "c", "a" ],
    //   [ "a", "c", "b" ],
    //   [ "c", "a", "b" ],
    //   [ "c", "b", "a" ]
    // ]
    printVector(permutations(std::vector<std::string>{"red", "blue"})); // ->
    // [
    //   [ "red", "blue" ],
    //   [ "blue", "red" ]
    // ]
    printVector(permutations(std::vector<std::string>{"8", "2", "1", "4"})); // ->
    // [
    //   [ "8", "2", "1", "4" ], [ "2", "8", "1", "4" ],
    //   [ "2", "1", "8", "4" ], [ "2", "1", "4", "8" ],
    //   [ "8", "1", "2", "4" ], [ "1", "8", "2", "4" ],
    //   [ "1", "2", "8", "4" ], [ "1", "2", "4", "8" ],
    //   [ "8", "1", "4", "2" ], [ "1", "8", "4", "2" ],
    //   [ "1", "4", "8", "2" ], [ "1", "4", "2", "8" ],
    //   [ "8", "2", "4", "1" ], [ "2", "8", "4", "1" ],
    //   [ "2", "4", "8", "1" ], [ "2", "4", "1", "8" ],
    //   [ "8", "4", "2", "1" ], [ "4", "8", "2", "1" ],
    //   [ "4", "2", "8", "1" ], [ "4", "2", "1", "8" ],
    //   [ "8", "4", "1", "2" ], [ "4", "8", "1", "2" ],
    //   [ "4", "1", "8", "2" ], [ "4", "1", "2", "8" ]
    // ]
    printVector(permutations(std::vector<std::string>{})); // ->
                                                           // [
                                                           //  [ ]
                                                           // ]
    return 0;
}

File: javascript_version/exhaustive_recursion/permutations/permutations.js
// permutations
// Write a function, permutations, that takes in a list an argument. The function should 
// return a 2D list where each sublist represents one of the possible permutations of the list.

// The sublists may be returned in any order.

// You may assume that the input list contains unique items.


const permutations = (items) => {
    // todo
    if (items.length === 0) { return [[]]; }
    return _permutations(items);
};
const _permutations = (items) => {
    if (items.length === 0) { return [[]]; }
    let first = items[0];
    let result = [];
    for (let perm of _permutations(items.slice(1))) {
        for (let i = 0; i <= perm.length; i++) {
            result.push([...perm.slice(0, i), first, ...perm.slice(i)]);
        }
    }
    return result;
};



console.log(permutations(['a', 'b', 'c'])); // -> 
// [
//   [ 'a', 'b', 'c' ], 
//   [ 'b', 'a', 'c' ], 
//   [ 'b', 'c', 'a' ], 
//   [ 'a', 'c', 'b' ], 
//   [ 'c', 'a', 'b' ], 
//   [ 'c', 'b', 'a' ] 
// ] 
console.log(permutations(['red', 'blue'])); // ->
// [ 
//   [ 'red', 'blue' ], 
//   [ 'blue', 'red' ] 
// ]
console.log(permutations([8, 2, 1, 4])); // ->
// [ 
//   [ 8, 2, 1, 4 ], [ 2, 8, 1, 4 ], 
//   [ 2, 1, 8, 4 ], [ 2, 1, 4, 8 ], 
//   [ 8, 1, 2, 4 ], [ 1, 8, 2, 4 ], 
//   [ 1, 2, 8, 4 ], [ 1, 2, 4, 8 ], 
//   [ 8, 1, 4, 2 ], [ 1, 8, 4, 2 ], 
//   [ 1, 4, 8, 2 ], [ 1, 4, 2, 8 ], 
//   [ 8, 2, 4, 1 ], [ 2, 8, 4, 1 ], 
//   [ 2, 4, 8, 1 ], [ 2, 4, 1, 8 ], 
//   [ 8, 4, 2, 1 ], [ 4, 8, 2, 1 ], 
//   [ 4, 2, 8, 1 ], [ 4, 2, 1, 8 ], 
//   [ 8, 4, 1, 2 ], [ 4, 8, 1, 2 ], 
//   [ 4, 1, 8, 2 ], [ 4, 1, 2, 8 ] 
// ] 
console.log(permutations([])); // ->
// [
//  [ ]
// ]